fix(sidecar): make it respect node status on sync

dkropachev opened the pull request:
**Description of your changes:**
Make sidecar controller respect scylla state on decommission

**Which issue is resolved by this Pull Request:**
Resolves https://github.com/scylladb/scylla-operator/issues/223
Resolves https://github.com/scylladb/scylla-operator/issues/243

**Checklist:**
- [ ] ~~Documentation has been updated, if necessary.~~
- [x] Image has been built (`make docker-build`) on the last commit.

zimnx commented on pkg/controllers/sidecar/sync.go:
```diff
@@ -11,32 +11,53 @@ import (
+				return true, errors.Wrapf(err, "error on getting node operation mode after decommission")
+			}
+			// Confirm memberService has been decommissioned
+			if !opMode.IsDecommisioned() {
```
In case when decomission happens immediately this will trigger unnecessary error which changes nothing. Let's remove this conditon.

dkropachev replied:
removed

zimnx commented on pkg/controllers/sidecar/sync.go:
```diff
@@ -11,32 +11,53 @@ import (
+				return true, errors.Wrapf(err, "error during decommission, operation mode: %s", opMode)
+			}
+		} else if opMode.IsJoining() {
+			// If node is joining we need to wait till it is joined (IsNormal() == True) and then decommission it
```
```suggestion
			// If node is joining we need to wait till it reaches Normal state and then decommission it
```

dkropachev replied:
fixed

zimnx commented on pkg/controllers/sidecar/sync.go:
```diff
@@ -11,32 +11,53 @@ import (
+			return true, nil
+		} else if opMode.IsDecommisioned() {
+			mc.logger.Info(ctx, "Node is decommissioned. Skip decommission")
+			// Do nothing if it is already decommissioned
```
```suggestion
```
This comment content is already in log message

dkropachev replied:
removed

zimnx commented on pkg/controllers/sidecar/sync.go:
```diff
@@ -11,32 +11,46 @@ import (
-		// Confirm memberService has been decommissioned
-		if opMode, err := mc.scyllaClient.OperationMode(ctx, host.String()); err != nil || !opMode.IsDecommisioned() {
-			return errors.Wrapf(err, "error during decommission, operation mode: %s", opMode)
+		if opMode.IsLeaving() {
```
5 branches, let's use switch statement:
```
switch opMode {
case OperationalModeNormal:
// 
case OperationalModeDecommissioned:
// etc
```

dkropachev replied:
done

dkropachev commented:
@zimnx , Addressed all comments, found once case which was not covered - when node is drained.

Current implementation throws error that node needs to be restarted.
It would be really nice to silently restart node and proceed with decommissioning it automatically, but I don't know good way to do that.

zimnx commented on pkg/controllers/sidecar/sync.go:
```diff
@@ -2,41 +2,67 @@ package sidecar
+			case scyllaclient.OperationalModeNormal:
+				// Decommission node only if it is in normal mode
+				if decommissionErr := mc.scyllaClient.Decommission(ctx, host.String()); decommissionErr != nil {
+					// Decommission is long running task, so request fails due to the timeout in most cases
```
Shouldn't Scyla enter Leaving state after first request? Condition below suggest it does so this branch shouldn't be picked up on retry.
If next reconcilation happens too fast, we have control over it. 
https://github.com/scylladb/scylla-operator/blob/bc214d9b9dbb7462bbd436ae393e288d50fc386a/vendor/sigs.k8s.io/controller-runtime/pkg/reconcile/reconcile.go#L32

zimnx replied:
We can also check if timeout happened or it was some other failure and react based on error value

dkropachev replied:
I did not find that it does so.

Problem here that retrying is done on scyllaclient level, so sync is not getting 'control' back from mc.scyllaClient.Decommission(ctx, host.String()) up until scyllaclient exhausted all retrys.

Even if we get scyllaclient fixed, decommission can take unpredictable amount of time, so best thing to do here would be relaying on node status, it if is in decommissioned or decommissining state, then we got what we wanted.
Additionaly we cover all possible cases that could happen due to the race condition with nodetool commands.

dkropachev commented:
Decommissioning looks great now:

```
{"L":"ERROR","T":"2020-11-09T11:41:15.209Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"_c0slUtbQDyfQp0dAhKNkg"}
{"L":"ERROR","T":"2020-11-09T11:41:25.204Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"hc39rM-VSX25NcBoX1H3oQ"}
{"L":"ERROR","T":"2020-11-09T11:41:35.201Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"Tq_zvjPzQSeKPIwfXHI4_Q"}
{"L":"ERROR","T":"2020-11-09T11:41:45.207Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"r_ozJfGdSQ-XyGZ402E6JA"}
INFO  2020-11-09 11:44:25,973 [shard 0] storage_service - DECOMMISSIONING: starts
{"L":"ERROR","T":"2020-11-09T11:44:35.230Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"lmrH8lGcRsOisG55oxgGTQ"}
{"L":"DEBUG","T":"2020-11-09T11:44:40.973Z","N":"member_controller.scylla_client","M":"HTTP","host":"172.17.0.9:10001","method":"POST","uri":"/storage_service/decommission","duration":"15000ms","_trace_id":"uroC3bd2Tj2EWlS9GOe0PA"}
{"L":"ERROR","T":"2020-11-09T11:44:45.206Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"GdYZbg_ITniTGMck590HOA"}
{"L":"ERROR","T":"2020-11-09T11:44:55.202Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"BjeGZqe-SjCf4fMFJNc-mQ"}
INFO  2020-11-09 11:44:56,119 [shard 0] storage_service - DECOMMISSIONING: unbootstrap starts
{"L":"ERROR","T":"2020-11-09T11:45:05.199Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"3-dQNTHUQJS4uu1MzswPKA"}
{"L":"ERROR","T":"2020-11-09T11:45:15.204Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"Lyw5zhKETS-S0SKZ1EdXEA"}
{"L":"ERROR","T":"2020-11-09T11:45:25.200Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"Xeq2SVL3RbCgRWauBFNAmw"}
INFO  2020-11-09 11:45:26,560 [shard 0] storage_service - DECOMMISSIONING: unbootstrap done
INFO  2020-11-09 11:45:26,563 [shard 0] storage_service - DECOMMISSIONING: shutdown rpc and cql server done
INFO  2020-11-09 11:45:26,563 [shard 0] storage_service - DECOMMISSIONING: stop batchlog_manager done
INFO  2020-11-09 11:45:26,564 [shard 0] storage_service - DECOMMISSIONING: stop_gossiping done
INFO  2020-11-09 11:45:26,570 [shard 0] storage_service - DECOMMISSIONING: stop messaging_service done
INFO  2020-11-09 11:45:26,657 [shard 0] storage_service - DECOMMISSIONING: set_bootstrap_state done
INFO  2020-11-09 11:45:26,657 [shard 0] storage_service - DECOMMISSIONED: 
INFO  2020-11-09 11:45:26,657 [shard 0] storage_service - DECOMMISSIONING: done
{"L":"ERROR","T":"2020-11-09T11:45:35.200Z","N":"member_controller","M":"Readiness check failed, node not ready","_trace_id":"I2z-i6i2S2adLH3WdtGUuA"}
{"L":"INFO","T":"2020-11-09T11:45:38.597Z","N":"member_controller","M":"Node is decommissioned. Skip decommission","_trace_id":"P4nfsZ_rSJOyEQYLRkOtfQ"}
{"L":"DEBUG","T":"2020-11-09T11:45:38.615Z","N":"member_controller","M":"Node is already decommissioned","_trace_id":"2kYkwavUTJCGjbF1i0ZMaA"}
rpc error: code = Unknown desc = Error: No such container: de47028b6b9ff6dbba310684a6019eae08ef283cf328fc4b39a1c48590b6b9d9

```

But I still would like to restart scylla on DRAINED state, could you bump me to the right way to do that from sidecar.

zimnx commented on pkg/scyllaclient/client.go:
```diff
@@ -158,7 +158,9 @@ func (c *Client) Status(ctx context.Context, host string) (NodeStatusInfoSlice,
 func (c *Client) Decommission(ctx context.Context, host string) error {
-	_, err := c.scyllaOps.StorageServiceDecommissionPost(&scyllaOperations.StorageServiceDecommissionPostParams{Context: forceHost(ctx, host)})
+	queryCtx := forceHost(ctx, host)
+	queryCtx = noRetry(queryCtx)
```
Please add a comment explaining why retry is disabled - unfortunate 500 response.

dkropachev replied:
added

dkropachev commented:
One more case discovered - https://github.com/scylladb/scylla-operator/issues/243

Current revision covers following cases:
decomissionning
decomissionned
draininig
drained
startingup

[reduce_logs.zip](https://github.com/scylladb/scylla-operator/files/5523048/reduce_logs.zip)

zimnx commented on pkg/controllers/sidecar/sync.go:
```diff
@@ -2,41 +2,82 @@ package sidecar
 	"context"
-
-	"github.com/scylladb/scylla-operator/pkg/util/network"
+	"github.com/scylladb/scylla-operator/pkg/scyllaclient"
```
please run `goimports` on every file you've changed.

dkropachev replied:
fixed

zimnx reviewed (changes requested):
one comment inline + add "Fixed #243" to your commit msg

dkropachev commented:
> one comment inline + add "Fixed #243" to your commit msg

added
